src/robot/conf/gatherfailed.py

```diff
@@ -32,6 +32,21 @@ def visit_keyword(self, kw):
         pass
 
 
+class GatherFailedSuites(SuiteVisitor):
+
+    def __init__(self, critical_tags, non_critical_tags):
+        self.suites = set()
+        self._criticality = Criticality(critical_tags,
+                                        non_critical_tags)
```

Comment:
This ought to work, but it might be cleaner to call `suite.set_crticality()` when `start_suite` is called for the first time. It would especially simplify `visit_test` implementation.